unbricking.md

```diff
@@ -0,0 +1,346 @@
+# G-010S-A Unbricking
+
+If anything goes wrong during an update the SFP module can be unbricked via
+netconsole or a serial console.
+
+## Netconsole Compatibility
+
+Netconsole does not work in all circumstances. It might come down to specific
+hardware combinations:
+
+| SFP Module Model  | SFP Host                  | Netconsole working |
+|-------------------|---------------------------|--------------------|
+| 3FE46541AAAA      | BCM57810S                 | Unstable           |
+| 3FE46541AAAA      | Media Converter           | Unstable           |
+| 3FE46542AABA      | MC220L                    | Y                  |
+| 3FE46541AADA      | 10Gtek WG-33-1GX1GT-SFP   | Y                  |
+| 3FE46541AADA      | Netgear GS752TPV2         | Y                  |
+| 3FE46541AADA      | CSR305                    | ?                  |
+
+
+## Preparation / Unlock Bootloader
+<a id="preparation"></a>
+
+By default the bootloader is locked and will not display a serial console / will
+not allow interrupting the boot. Netconsole is also disabled by default.
+
+After enabling any of the below make sure to check that ```bootdelay``` is set
+to a suitable value (i.e., a value greater than 0).
+
+### Enabling Serial Console and NetConsole
+
+WARNING: Opening the serial console could cause the "pin 6" issue ([known
+compatibility](README.md#compatibility)).
+
+To unlock the bootloader and enable serial and netconsole while the SFP module
+still working:
+```
+ONTUSER@SFP:~# serial open
+disabled serial and enable present pin
+```
+
+To disable the serial console again use:
+```
+ONTUSER@SFP:~# serial close
+disabled serial and enable present pin
+```
+
+#### Configure NetConsole Host
+
+To also start netconsole you need to set the target host for the console
+input/output:
+```
+ONTUSER@SFP:~# fw_setenv ncip 192.168.1.1
+```
+Replace the IP address with the right value for your network. It must be an
+address that is reachable directly by the SFP as no gateway is configured by
+default. The IP needs to respond to a ping at the time the SFP boots
+to start the netconsole.
+
+### Enabling NetConsole Only
+
+To enable NetConsole only use:
+```
+fw_setenv preboot 'gpio clear 100; run start_netconsole'
+fw_setenv asc0 0
+```
+
+Also configure the NetConsole host as above.
+
+To disable the NetConsole again use:
+```
+ONTUSER@SFP:~# serial close
+disabled serial and enable present pin
+```
+
+## U-boot Connection Using Netconsole
+
+Warning: Netconsole has not been consistently working for various setups. One
+possible factor may be if the SFP module is in a switch or media converter as
+opposed to in an SFP cage on the router. The exact u-boot version may be another.
```

Comment:
Now we can remove the reference to the u-boot version.

I would just keep the `Warning: Netconsole has not been consistently working for various setups.` and a link to the compatibility table.
There is no need to add any speculation here, we will add something if we get to a proper conclusion.

Reply:
What got the netconsole working in the end? Is it just the issue with the cold boot?

Reply:
With the BCM57810S, it often work on a cold boot (and even better on a reset from u-boot), but almost never on a reboot.
With the CRS305, with additional config changes, it can work on a reboot, but not on a cold boot (so it's almost useless, especially since it breaks the Module Present once booted).

Since I can't get the UART traces at the same time, it's difficult to tell what happens.
Each device seems to have its own definition of "link up" :
 - The CRS305 is unable to receive anything on a cold boot (maybe the ARP is sent too early, adding a sleep an another ping doesn't seem to help, but maybe it doesn't work, I can't see without  the UART).
 - The BCM57810S is able to receive the ARP with the link down (both on Linux side, and from the LEDs).

Reply:
Maybe it's just u-boot which doesn't properly reset all registers, and/or doesn't wait enough time in the interface setup code.
But https://github.com/minhng99/u-boot_lantiq_falcon/blob/master/drivers/net/falcon/falcon.c only has SGMII, not SERDES, so it wouldn't be easy to try to understand what u-boot does.